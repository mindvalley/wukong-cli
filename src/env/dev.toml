debug = true

[api]
url = "http://localhost:4000/api"

[okta]
client_id = "0oakfxaegyAV5JDD5357"
