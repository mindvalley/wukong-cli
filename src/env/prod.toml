debug = false

[api]
url = ""

[okta]
client_id = ""
