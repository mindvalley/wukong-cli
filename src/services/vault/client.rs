use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize)]
pub struct Auth {
    pub client_token: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Login {
    pub auth: Auth,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FetchSecretsData {
    pub data: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FetchSecrets {
    pub data: FetchSecretsData,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VerifyTokenData {
    pub expire_time: String,
    pub issue_time: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VerifyToken {
    pub data: VerifyTokenData,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateSecret {
    wrap_info: Option<String>,
    warnings: Option<String>,
    auth: Option<Auth>,
}

pub struct VaultClient {
    client: reqwest::Client,
    base_url: String,
}

impl VaultClient {
    pub const LOGIN: &str = "/v1/auth/okta/login";
    pub const VERIFY_TOKEN: &str = "/v1/auth/token/lookup-self";
    pub const FETCH_SECRETS: &str = "/v1/secret/data";
    pub const UPDATE_SECRET: &str = "/v1/secret/data";

    pub fn new() -> Self {
        let client = reqwest::Client::new();
        #[cfg(all(feature = "prod"))]
        let base_url = "https://bunker.mindvalley.com".to_string();

        #[cfg(not(feature = "prod"))]
        let base_url = match std::env::var("WUKONG_DEV_VAULT_API_URL") {
            Ok(vault_api_url) => vault_api_url,
            Err(_) => "https://bunker.mindvalley.dev:8200".to_string(),
        };

        Self { client, base_url }
    }

    #[cfg(test)]
    pub fn with_base_url(mut self, base_url: String) -> Self {
        self.base_url = base_url;
        self
    }

    pub async fn login(
        &self,
        email: &str,
        password: &str,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let url = format!("{}{}/{}", self.base_url, Self::LOGIN, email);

        let response = self
            .client
            .post(url)
            .form(&[("password", password)])
            .send()
            .await?;

        Ok(response)
    }

    pub async fn fetch_secrets(
        &self,
        api_token: &str,
        path: &str,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let url = format!("{}{}/{}", self.base_url, Self::FETCH_SECRETS, path);

        let response = self
            .client
            .get(url)
            .header("X-Vault-Token", api_token)
            .send()
            .await?;

        println!("response: {:?}", response);

        Ok(response)
    }

    pub async fn verify_token(&self, api_token: &str) -> Result<reqwest::Response, reqwest::Error> {
        let url = format!("{}{}", self.base_url, Self::VERIFY_TOKEN);

        let response = self
            .client
            .get(url)
            .header("X-Vault-Token", api_token)
            .send()
            .await?;

        Ok(response)
    }

    pub async fn update_secret(
        &self,
        api_token: &str,
        path: &str,
        data: &HashMap<&str, &str>,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let url = format!("{}{}/{}", self.base_url, Self::UPDATE_SECRET, path);

        // Update the secret with updated value:
        let mut secret_data = HashMap::new();
        secret_data.insert("data", data);

        let response = self
            .client
            .post(url)
            .header("X-Vault-Token", api_token)
            .json(&secret_data)
            .send()
            .await?;

        Ok(response)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use httpmock::prelude::*;

    #[tokio::test]
    async fn test_login() {
        let server = MockServer::start();

        let email = "test@example.com";
        let password = "test_password";

        let api_resp = r#"
            {
              "auth": {
                "client_token": "test_token"
                }
            }"#;

        let mock_server = server.mock(|when, then| {
            when.method(POST)
                .path_contains(email)
                .path_contains(VaultClient::LOGIN)
                .body(format!("password={}", password));
            then.status(200)
                .header("content-type", "application/json; charset=UTF-8")
                .body(api_resp);
        });

        let vault_client = VaultClient::new().with_base_url(server.base_url());
        let response = vault_client.login(email, password).await;

        mock_server.assert();
        assert!(response.is_ok());

        let response = response.unwrap();
        assert_eq!(response.status(), 200);

        let login_data = response.json::<Login>().await.unwrap();
        assert_eq!(login_data.auth.client_token, "test_token");
    }

    #[tokio::test]
    async fn test_login_failed_with_bad_credentials() {
        let server = MockServer::start();

        let email = "test@example.com";
        let password = "wrong_password";

        let api_resp = r#"
            {
              "errors": ["Okta auth failed"]
            }"#;

        let mock_server = server.mock(|when, then| {
            when.method(POST)
                .path_contains(VaultClient::LOGIN)
                .path_contains(email)
                .body(format!("password={}", password));
            then.status(400)
                .header("content-type", "application/json; charset=UTF-8")
                .body(api_resp);
        });

        let vault_client = VaultClient::new().with_base_url(server.base_url());
        let response = vault_client.login(email, password).await;

        mock_server.assert();

        let response = response.unwrap();
        assert_eq!(response.status(), 400);
    }

    #[tokio::test]
    async fn test_verify_token() {
        let server = MockServer::start();

        let api_token = "secret_token";

        let api_resp = r#"
            {
              "data": {
                "expire_time": "2019-12-10T10:10:10.000000Z",
                "issue_time": "2019-10-10T10:10:10.000000Z"
                }
            }"#;

        let mock_server = server.mock(|when, then| {
            when.method(GET)
                .path_contains(VaultClient::VERIFY_TOKEN)
                .header("X-Vault-Token", api_token);
            then.status(200)
                .header("content-type", "application/json; charset=UTF-8")
                .body(api_resp);
        });

        let vault_client = VaultClient::new().with_base_url(server.base_url());
        let response = vault_client.verify_token(api_token).await;

        mock_server.assert();
        assert!(response.is_ok());

        let response = response.unwrap();
        assert_eq!(response.status(), 200);

        let verify_token = response.json::<VerifyToken>().await.unwrap();

        assert!(
            !verify_token.data.expire_time.is_empty(),
            "Value should not be None"
        );
    }

    #[tokio::test]
    async fn test_fetch_secrets() {
        let server = MockServer::start();

        let api_token = "test_token";
        let path = "devenv/test";

        let api_resp = r#"
            {
              "data": {
                "data": {
                    "test2": "secret_token"
                    }
                }
            }"#;

        let mock_server = server.mock(|when, then| {
            when.method(GET)
                .path_contains("/v1/secret/data")
                .path_contains(path)
                .header("X-Vault-Token", api_token);
            then.status(200)
                .header("content-type", "application/json; charset=UTF-8")
                .body(api_resp);
        });

        let vault_client = VaultClient::new().with_base_url(server.base_url());
        let response = vault_client.fetch_secrets(api_token, path).await;

        mock_server.assert();
        assert!(response.is_ok());

        let response = response.unwrap();
        assert_eq!(response.status(), 200);

        let secrets = response.json::<FetchSecrets>().await.unwrap();

        let data = secrets.data.data.get("test2").unwrap();
        assert_eq!(data, "secret_token");
    }

    #[tokio::test]
    async fn test_update_secret() {
        let server = MockServer::start();

        let api_token = "test_token";
        let path = "devenv/test";
        let mut update_data = HashMap::new();
        update_data.insert("test", "test4");

        let api_resp = r#"
            {
              "data": {
                "test": "test4"
                }
            }"#;

        let mock_server = server.mock(|when, then| {
            when.method(POST)
                .path_contains(VaultClient::UPDATE_SECRET)
                .body(format!(r#"{{"data":{{"{}":"{}"}}}}"#, "test", "test4"))
                .header("X-Vault-Token", api_token);
            then.status(200)
                .header("content-type", "application/json; charset=UTF-8")
                .body(api_resp);
        });

        let vault_client = VaultClient::new().with_base_url(server.base_url());

        let response = vault_client
            .update_secret(api_token, path, &update_data)
            .await;

        mock_server.assert();
        assert!(response.is_ok());

        let response = response.unwrap();
        assert_eq!(response.status(), 200);
    }
}
