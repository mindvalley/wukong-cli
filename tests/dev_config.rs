mod common;
use aion::*;
use assert_fs::{
    fixture::ChildPath,
    prelude::{FileTouch, FileWriteStr, PathChild},
};
use httpmock::{Method::GET, MockServer};
use serial_test::serial;
use std::env;
use wukong::services::vault::client::VaultClient;

fn setup() -> (assert_fs::TempDir, assert_fs::TempDir) {
    let wk_temp = assert_fs::TempDir::new().unwrap();
    let elixir_temp = assert_fs::TempDir::new().unwrap();

    env::set_current_dir(elixir_temp.path().to_str().unwrap()).unwrap();

    (wk_temp, elixir_temp)
}

fn teardown(wk_temp: assert_fs::TempDir, elixir_temp: assert_fs::TempDir) {
    wk_temp.close().unwrap();
    elixir_temp.close().unwrap();
}

fn verify_token_mock(server: &MockServer) -> httpmock::Mock {
    let verify_token_api_resp = r#"
        {
          "data": {
            "expire_time": "2019-12-10T10:10:10.000000Z",
            "issue_time": "2019-10-10T10:10:10.000000Z"
            }
        }"#;

    server.mock(|when, then| {
        when.method(GET)
            .path_contains(VaultClient::VERIFY_TOKEN)
            .header("X-Vault-Token", "valid_vault_api_token");
        then.status(200)
            .header("content-type", "application/json; charset=UTF-8")
            .body(verify_token_api_resp);
    })
}

fn get_secret_mock<'a>(server: &'a MockServer, custom_data: Option<&'a str>) -> httpmock::Mock<'a> {
    let data = custom_data.unwrap_or_else(|| r#"
        {
          "b.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
          "c.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5"
        }"#);

    let secret_api_resp = format!(
        r#"{{
          "data": {{
            "data": {},
            "metadata": {{
              "created_time": "2015-02-22T02:24:06.945319214Z",
              "custom_metadata": {{
                "owner": "xxx",
                "mission_critical": "false"
              }},
              "deletion_time": "",
              "destroyed": false,
              "version": 2
            }}
          }}
        }}"#,
        data
    );

    server.mock(|when, then| {
        when.method(GET).path_contains(VaultClient::FETCH_SECRETS);
        then.status(200)
            .header("content-type", "application/json; charset=UTF-8")
            .body(secret_api_resp);
    })
}

fn mock_user_config(wk_temp: &assert_fs::TempDir, server_url: String) -> ChildPath {
    let wk_config_file = wk_temp.child("config.toml");
    wk_config_file.touch().unwrap();

    wk_config_file
        .write_str(
            format!(
                r#"
                    [core]
                    application = "valid-application"
                    wukong_api_url = "{}"
                    okta_client_id = "valid-okta-client-id"

                    [vault]
                    api_token = "valid_vault_api_token"
                    expiry_time = "2027-06-09T08:51:19.032792+00:00"

                    [auth]
                    account = "test@email.com"
                    subject = "subject"
                    id_token = "id_token"
                    access_token = "access_token"
                    expiry_time = "{}"
                    refresh_token = "refresh_token"
                "#,
                server_url,
                2.days().from_now().to_rfc3339()
            )
            .as_str(),
        )
        .unwrap();

    wk_config_file
}

#[test]
fn test_wukong_dev_config_help() {
    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("help")
        .assert()
        .success();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stdout).unwrap());
}

#[test]
#[serial]
fn test_wukong_dev_config_diff_success() {
    let (wk_temp, elixir_temp) = setup();
    let server = MockServer::start();

    let verify_token_mock = verify_token_mock(&server);
    let secret_data_mock = get_secret_mock(&server, None);
    let wk_config_file = mock_user_config(&wk_temp, server.base_url());

    let dev_config_file = elixir_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();
    let dev_config_secret_file_c = elixir_temp.child("config/c.secret.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
                use Mix.Config

                System.get_env("API_KEY")
                System.fetch_env("API_SECRET")
                System.fetch_env!("API_TOKEN")

                # wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#c.secret.exs
                File.exists?("c.secret.exs") && import_config "c.secret.exs"

                test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

                # Use Jason for JSON parsing in Phoenix
                config :phoenix, :json_library, Jason
            "#,
        )
        .unwrap();

    dev_config_secret_file_c
        .write_str(
            r#"
use Mix.Config

# Configure your database.
config :academy, Academy.Repo,
  adapter: Ecto.Adapters.Postgres,
  username: "postgres",
  passwords: "",
  database: "academy_core_devss",
  pool_size: 10
            "#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("diff")
        .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
        .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
        .assert()
        .success();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stdout).unwrap());

    verify_token_mock.assert();
    secret_data_mock.assert();

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_diff_when_secret_key_not_found_from_bunker() {
    let (wk_temp, elixir_temp) = setup();
    let server = MockServer::start();

    let verify_token_mock = verify_token_mock(&server);
    let secret_data_mock = get_secret_mock(
        &server,
        Some(
            r#"
        {
          "a.secret.exs": "test"
        }"#,
        ),
    );
    let wk_config_file = mock_user_config(&wk_temp, server.base_url());

    let dev_config_file = elixir_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();
    let dev_config_secret_file = elixir_temp.child("config/c.secret.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
                use Mix.Config

                System.get_env("API_KEY")
                System.fetch_env("API_SECRET")
                System.fetch_env!("API_TOKEN")

                # wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#c.secret.exs
                File.exists?("c.secret.exs") && import_config "c.secret.exs"

                test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

                # Use Jason for JSON parsing in Phoenix
                config :phoenix, :json_library, Jason
            "#,
        )
        .unwrap();

    dev_config_secret_file
        .write_str(
            r#"use Mix.Config

config :application, Application.Repo"#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("diff")
        .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
        .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
        .assert()
        .failure();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());

    verify_token_mock.assert();
    secret_data_mock.assert();

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_diff_when_secret_file_not_found() {
    let (wk_temp, elixir_temp) = setup();
    let server = MockServer::start();

    let verify_token_mock = verify_token_mock(&server);
    let secret_data_mock = get_secret_mock(
        &server,
        Some(
            r#"
        {
              "b.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
              "c.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo"
            }"#,
        ),
    );
    let wk_config_file = mock_user_config(&wk_temp, server.base_url());
    let dev_config_file = elixir_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
                use Mix.Config

                System.get_env("API_KEY")
                System.fetch_env("API_SECRET")
                System.fetch_env!("API_TOKEN")

                # wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#c.secret.exs
                File.exists?("c.secret.exs") && import_config "c.secret.exs"

                test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

                # Use Jason for JSON parsing in Phoenix
                config :phoenix, :json_library, Jason
            "#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("diff")
        .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
        .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
        .assert()
        .failure();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());

    verify_token_mock.assert();
    secret_data_mock.assert();

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_diff_when_no_changes_found() {
    let (wk_temp, elixir_temp) = setup();
    let server = MockServer::start();

    let verify_token_mock = verify_token_mock(&server);
    let secret_data_mock = get_secret_mock(
        &server,
        Some(
            r#"
        {
              "b.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
              "c.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo"
            }"#,
        ),
    );
    let wk_config_file = mock_user_config(&wk_temp, server.base_url());

    let dev_config_file = elixir_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();
    let dev_config_secret_file = elixir_temp.child("config/c.secret.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
                use Mix.Config

                System.get_env("API_KEY")
                System.fetch_env("API_SECRET")
                System.fetch_env!("API_TOKEN")

                # wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#c.secret.exs
                File.exists?("c.secret.exs") && import_config "c.secret.exs"

                test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

                # Use Jason for JSON parsing in Phoenix
                config :phoenix, :json_library, Jason
            "#,
        )
        .unwrap();

    dev_config_secret_file
        .write_str(
            r#"use Mix.Config

config :application, Application.Repo"#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("diff")
        .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
        .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
        .assert()
        .success();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stdout).unwrap());

    verify_token_mock.assert();
    secret_data_mock.assert();

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_diff_when_config_not_found() {
    let (wk_temp, elixir_temp) = setup();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("diff")
        .assert()
        .failure();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_pull_success() {
    let (wk_temp, elixir_temp) = setup();
    let server = MockServer::start();

    let secret_data_mock = get_secret_mock(
        &server,
        Some(
            r#"
        {
              "a.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
              "b.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
              "c.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5"
            }"#,
        ),
    );
    let verify_token_mock = verify_token_mock(&server);
    let wk_config_file = mock_user_config(&wk_temp, server.base_url());

    wk_config_file.touch().unwrap();

    wk_config_file
        .write_str(
            format!(
                r#"
[core]
application = "valid-application"
wukong_api_url = "{}"
okta_client_id = "valid-okta-client-id"

[vault]
api_token = "valid_vault_api_token"
expiry_time = "2027-06-09T08:51:19.032792+00:00"

[auth]
account = "test@email.com"
subject = "subject"
id_token = "id_token"
access_token = "access_token"
expiry_time = "{}"
refresh_token = "refresh_token"
    "#,
                server.base_url(),
                2.days().from_now().to_rfc3339()
            )
            .as_str(),
        )
        .unwrap();

    let dev_config_file = elixir_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
use Mix.Config

System.get_env("API_KEY")
System.fetch_env("API_SECRET")
System.fetch_env!("API_TOKEN")

# wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#a.secret.exs
import_config "config/a.secret.exs"

# wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#b.secret.exs
if File.exists?("config/b.secret.exs") do
  import_config "config/b.secret.exs"
end

# wukong.mindvalley.dev/config-secrets-location: vault:secret/mv/tech/app/dev#c.secret.exs
File.exists?("config/c.secret.exs") && import_config "config/c.secret.exs"

test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

# Use Jason for JSON parsing in Phoenix
config :phoenix, :json_library, Jason
    "#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("pull")
        .arg(elixir_temp.path().to_str().unwrap())
        .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
        .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
        .assert()
        .success();

    let output = cmd.get_output();

    insta::with_settings!({filters => vec![
        (format!("{}", elixir_temp.path().to_str().unwrap()).as_str(), "[TEMP_DIR]"),
    ]}, {
        insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());
    });

    verify_token_mock.assert();
    secret_data_mock.assert();

    teardown(wk_temp, elixir_temp)
}

#[test]
#[serial]
fn test_wukong_dev_config_lint_on_main_config_success() {
    let wk_temp = assert_fs::TempDir::new().unwrap();
    let main_config_file = wk_temp.child("config/config.exs");
    main_config_file.touch().unwrap();

    main_config_file
        .write_str(
            r#"
use Mix.Config

System.get_env("API_KEY")
System.fetch_env("API_SECRET")
System.fetch_env!("API_TOKEN")

# invalid
import_config "config/dev.exs"

# valid
if File.exists?("config/dev.exs") do
  import_config "config/dev.exs"
end

# invalid
if File.exists?("config/a.exs") do
  import_config "config/b.exs"
end

# valid
File.exists?("config/dev.exs") && import_config "config/dev.exs"
# invalid
File.exists?("config/a.exs") && import_config "config/b.exs"

test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

# Use Jason for JSON parsing in Phoenix
config :phoenix, :json_library, Jason
    "#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("lint")
        .arg(wk_temp.path().to_str().unwrap())
        .assert()
        .failure();

    let output = cmd.get_output();

    insta::with_settings!({filters => vec![
        (format!("{}", wk_temp.path().to_str().unwrap()).as_str(), "[TEMP_DIR]"),
        (r"\d+(.\d+)ms|\d+(.\d+)s", "[DURATION]"),
    ]}, {
        insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());
    });

    wk_temp.close().unwrap();
}

#[test]
#[serial]
fn test_wukong_dev_config_lint_on_dev_config_success() {
    let wk_temp = assert_fs::TempDir::new().unwrap();
    let dev_config_file = wk_temp.child("config/dev.exs");
    dev_config_file.touch().unwrap();

    dev_config_file
        .write_str(
            r#"
use Mix.Config

System.get_env("API_KEY")
System.fetch_env("API_SECRET")
System.fetch_env!("API_TOKEN")

# invalid
import_config "config/dev.exs"

# valid
if File.exists?("config/dev.exs") do
  import_config "config/dev.exs"
end

# invalid
if File.exists?("config/a.exs") do
  import_config "config/b.exs"
end

# valid
File.exists?("config/dev.exs") && import_config "config/dev.exs"
# invalid
File.exists?("config/a.exs") && import_config "config/b.exs"

test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")

# Use Jason for JSON parsing in Phoenix
config :phoenix, :json_library, Jason
    "#,
        )
        .unwrap();

    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("config")
        .arg("lint")
        .arg(wk_temp.path().to_str().unwrap())
        .assert()
        .failure();

    let output = cmd.get_output();

    insta::with_settings!({filters => vec![
        (format!("{}", wk_temp.path().to_str().unwrap()).as_str(), "[TEMP_DIR]"),
        (r"\d+(.\d+)ms|\d+(.\d+)s", "[DURATION]"),
    ]}, {
        insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());
    });

    wk_temp.close().unwrap();
}
