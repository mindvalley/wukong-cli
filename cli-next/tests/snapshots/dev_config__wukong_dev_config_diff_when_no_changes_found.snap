---
source: tests/dev_config.rs
expression: "std::str::from_utf8(&output.stdout).unwrap()"
---
[36mcomparing local config vs remote config...[39m
The config file is already up to date with the Vault Bunker.

